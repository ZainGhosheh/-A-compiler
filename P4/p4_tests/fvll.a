# Multiply declared identifier
a : () -> int {
    x: int = 1;
    x: int = 2; # Error: Multiply declared identifier
    return x;
}

d : () -> int {
    y: int = 3;
    y: int = 4; # Error: Multiply declared identifier
    return y;
}

# Undeclared identifier
b : () -> int {
    return y; # Error: Undeclared identifier 'y'
}

e : () -> int {
    z: int = 5;
    return w; # Error: Undeclared identifier 'w'
}

# Invalid type in declaration
c : () -> void { # Error: Invalid type in declaration
    x: int = 1;
    return;
}

f : (param: void) -> int { # Error: Invalid type in declaration
    return 0;
}

g : () -> int {
    h: void; # Error: Invalid type in declaration
    return 0;
}

g : () -> int { # redefine function
    return 0;
}

c : custom {
    x: int;
    #y: string;
};

c22 : custom {
    x: int;
    y: void;
    #y: string;
};

c23 : custom {
    x: int;
    #y: string;
};

hello : () -> int {
    jjj: c22;
    kk: c23;
    kk->x = 1;
    kk->z = 1; # does not exist
    k: c24; # both class and member does not exist
    k->e = 2;
    x: immutable int;
    z: immutable void;
    return 0;
}

hello2 : () -> int {
    a: &int;
    #c: immutable a;
    b: &c23; # correct
    c: &a; # not allowed
    c: &void; # TODO
    maybe c means 4 otherwise 5;
    return 0;
}
