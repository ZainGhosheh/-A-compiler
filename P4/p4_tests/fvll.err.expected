FATAL [4,5]-[4,6]: Multiply declared identifier
FATAL [10,5]-[10,6]: Multiply declared identifier
FATAL [16,12]-[16,13]: Undeclared identifier
FATAL [21,12]-[21,13]: Undeclared identifier
FATAL [30,6]-[30,11]: Invalid type in declaration
FATAL [35,5]-[35,6]: Invalid type in declaration
FATAL [39,1]-[39,2]: Multiply declared identifier
FATAL [43,1]-[46,3]: Multiply declared identifier
FATAL [50,5]-[50,6]: Invalid type in declaration
FATAL [60,10]-[60,13]: Invalid type in declaration
FATAL [63,9]-[63,10]: Undeclared identifier
FATAL [64,8]-[64,11]: Invalid type in declaration
FATAL [65,5]-[65,6]: Undeclared identifier
FATAL [67,5]-[67,6]: Invalid type in declaration
FATAL [75,9]-[75,10]: Invalid type in declaration
FATAL [76,5]-[76,6]: Invalid type in declaration
ToDoError: implement me
